### ISSUE TITLE
Review serialization flow

### ISSUE THREAD

**Issue by rpadrela** [coherence/replication-server#201]
**Background**

Current serialization flow forces the replication-server to know about low level concepts such as streams and octets. See here [](https://github.com/coherence/replication-server/blob/5959829849f03140d6731b7a2917b8bf5e942b30/src/server/connection.go#L192).

The replication-server should only know about a few things such as ECS, live query, load balancer, etc. - all high level concepts. All of this is part of our domain and do not involve any details (serialization, streams, sockets, etc.).

Current serialization flow is being triggered at a pre-configured frequency and it asks brisk-protocol-go to send a stream. Brisk-protocol-go redirects the flow all the way up to the replication-server through an interface that the replication-server implements (see [here](https://github.com/coherence/replication-server/blob/5959829849f03140d6731b7a2917b8bf5e942b30/src/server/connection.go)), the server then redirects to protocol-serializer, ecs-sync-go, delta-ecs-go, etc., before returning back to brisk-protocol-go for the actual sending through the socket.

**Proposed change**

The function that is called at the pre-configured frequency could instead explicitly generate the update that needs to be pushed to clients by making use of the live query (live query would have to be moved out of the protocol-serializer - see [here](https://github.com/coherence/protocol-serializer/issues/94)), then, it would push those results through an interface for which the concrete implementation understands that updates need to be serialized through protocol-serializer, ecs-sync-go, delta-ecs-go, etc.

Flow-of-control wouldn't go back and forth between replication-server and brisk-protocol-go and the server wouldn't have to know about low level details (e.g. streams and octets).

**Expected**

- Investigate the feasibility of the suggested solution and the amount of work required

**Comment by rpadrela** [comment]
@cbrisebois, happy to know your thoughts on this.

**Comment by cbrisebois** [comment]
Yes, I generally agree. Brisk should not drive anything (this counts for the client too).  It should manage maintaining the connection via tend with ping / pong and then send a packet stream with called upon by replication-client-unity (pushed by the schema on the client) or replication-client-go (pushed by something new on the replication server when there is data to replicate)  I think that Brisk-protocol-go could also be a really poor name since it's not really part of Brisk but uses it.  I think a lot of the things BPG does can be moved up unto the replication server if that's what you're thinking.

**Comment by rpadrela** [comment]
At a high level I'd like to see something simple like this (pseudo-code):

```
func Tick() {
    update := generateUpdate()
    if update != nil {
           pushToClients(update)
    }
}
```

If I'm interested in how an update is generated, I'll "zoom in" on `generateUpdate()`. If I'm interested on how updates are pushed to clients, I'll "zoom in" on `pushToClients()`.

**Comment by cbrisebois** [comment]
Yes, I like this direction.  Update will be a byte array, right?  Or are you thinking a more complex object like a parts world?

**Comment by rpadrela** [comment]
> Yes, I like this direction. Update will be a byte array, right? Or are you thinking a more complex object like a parts world?

I was thinking of a more complex object which then gets transformed into a byte array through serialization. The `generateUpdate()` does not know what the update is for and does not know anything about serialization. All it does is generating an object that describes an update.
